RecoDataProducts/inc/CosmicTrack.hh

```diff
@@ -0,0 +1,333 @@
+#include "Mu2eUtilities/inc/BuildMatrixSums.hh"
+
+#include<vector>
+#include<bitset>
+
+using CLHEP::Hep3Vector;
+using namespace std;
+   //Struct To Hold Covarience Info:
+   struct TrackCov{
+   	std::vector<double> Covariance;
+	double sigA0;
+	double sigA0A1;
+        double sigA1A0;
+  	double sigA1;
+  	double sigB0;
+	double sigB0B1;
+        double sigB1B0;
+  	double sigB1;
+  	TrackCov();
+  	TrackCov(double siga0, double siga0a1,double siga1a0, double siga1, double sigb0,double sigb0b1, double sigb1b0, double sigb1) : sigA0(siga0), sigA0A1(siga0a1), sigA1A0(siga1a0), sigA1(siga1), sigB0(sigb0), sigB0B1(sigb0b1), sigB1B0(sigb1b0), sigB1(sigb1) {};
+   
+   };
+   //Struct To Hold Track Parameters
+   struct TrackParams{
+  	double A0;
+  	double A1;
+  	double B0;
+  	double B1;
+  	double T0;
+	double deltaA0;
+  	double deltaA1;
+  	double deltaB0;
+  	double deltaB1;
+  	double deltaT0;
+  	TrackParams();
+  	TrackParams(double a0, double a1, double  b0, double b1) : A0(a0), A1(a1), B0(b0) , B1(b1){};
+	TrackCov Covarience;
+	
+   };
+   //Struct to hold Coordinate System
+   struct TrackAxes{
+  	XYZVec _XDoublePrime;
+  	XYZVec _YDoublePrime;
+  	XYZVec _ZPrime;
+  	TrackAxes();
+  	TrackAxes(XYZVec X, XYZVec Y, XYZVec Z) : _XDoublePrime(X),_YDoublePrime(Y),_ZPrime(Z){};
+  	
+   };
+   //Struct to store a Track Equation (optional)
+   struct TrackEquation{
+   	XYZVec Pos;
+        XYZVec Dir;
+        TrackEquation();
+        TrackEquation(XYZVec P, XYZVec D) : Pos(P), Dir(D){};
+   };
+   //Struct to store Diagnostics associated with seed fit:
+   struct TrackSeedDiag{
+   	double FinalChiX;
+   	double FinalChiY;
+   	double FinalChiTot;
+   	
+   	double InitialChiX;
+   	double InitialChiY;
+   	double InitialChiTot;
+   	
+   	std::vector<double> InitialResidualsX;
+   	std::vector<double> InitialResidualsY;
+   	std::vector<double> FinalResidualsX;
+   	std::vector<double> FinalResidualsY;
+	 
+	std::vector<double> InitErrX;
+	std::vector<double> InitErrY;
+	std::vector<double> FinalErrX;
+	std::vector<double> FinalErrY; 
+	std::vector<double> FinalErrTot;
+	std::vector<double> InitErrTot;
+	TrackSeedDiag();
+
+   };
+   struct TrackDriftDiag{
+        double FinalChiX;
+   	double FinalChiY;
+   	double FinalChiTot;
+   	double NLL;
+   	
+   	std::vector<double> StartDOCAs;
+   	std::vector<double> FullFitEndDOCAs;
+	std::vector<double> GaussianEndDOCAs;
+	std::vector<double> TrueDOCAs;
+   	std::vector<double> StartTimeResiduals;
+   	std::vector<double> FullFitEndTimeResiduals;
+	std::vector<double> GaussianEndTimeResiduals;
+	std::vector<double> TrueTimeResiduals;
+	
+   	std::vector<double> RecoAmbigs;
+	std::vector<double> TrueAmbigs;
+
+   	std::vector<double> FinalResidualsX;
+   	std::vector<double> FinalResidualsY;
+	 
+	std::vector<double> FinalErrX;
+	std::vector<double> FinalErrY; 
+	std::vector<double> FinalErrTot;
+	
+   	TrackDriftDiag();
+   
+   };
+namespace mu2e {
+  
+  class CosmicTrack{
```

Comment:
Is there anything about this class that is specific to 'Cosmic'?  ie is it really just a straight track?
Diagnostic information should be kept separate from the actual data describing the fit.  Much of the content of the *Diag structs seems to be computable from the fit, do you need persistence for that or could it be function content.

Reply:
I liked the accessbility of these directly, the way this is done currently underpins the routine so it would take many weeks to re-do.....are there strong opinions on this?

Reply:
You can't put MC truth diagnostic info into a RecoDataProducts class.  The other embedded diagnostics can be kept if you really need them.  Eventually you will have to refactor this code.

Reply:
Ok, maybe I should move cosmic track to "data product" then and not "recodataproduct" the truth info is currently used for the analysis purposes not in the fitting routines, I need to think this through some more as to were is best to put the truth info. The truth info is only filled if mcdiag is set in the fitter.....this would not be the case if we were looking at data for example and then these truth members would just be empty....I thought that would be ok but I guess not...I'll rethink

Reply:
For now I have just removed the truth info. As this progresses, this class may change shape but for current purposes I think itis now ok.